File: backend/app/services/websocket.py
"""WebSocket fan-out for live dashboard updates.

Event types: cyclone.detected, cyclone.updated, prediction.issued,
alert.issued, alert.acknowledged, pipeline.status. See docs/api-contract.md §1.
"""

from datetime import UTC, datetime

from fastapi import WebSocket, WebSocketDisconnect


class ConnectionManager:
    def __init__(self) -> None:
        self.active: list[WebSocket] = []

    async def connect(self, websocket: WebSocket) -> None:
        await websocket.accept()
        self.active.append(websocket)

    def disconnect(self, websocket: WebSocket) -> None:
        if websocket in self.active:
            self.active.remove(websocket)

    async def broadcast(self, event_type: str, payload: dict) -> None:
        message = {
            "type": event_type,
            "timestamp": datetime.now(UTC).isoformat(),
            "payload": payload,
        }
        for connection in list(self.active):
            try:
                await connection.send_json(message)
            except (WebSocketDisconnect, RuntimeError):
                self.disconnect(connection)

    async def endpoint(self, websocket: WebSocket) -> None:
        await self.connect(websocket)
        try:
            while True:
                # Client messages are subscription hints; the server is push-driven.
                await websocket.receive_text()
        except WebSocketDisconnect:
            self.disconnect(websocket)


connection_manager = ConnectionManager()

File: frontend/src/services/websocket.js
/**
 * WebSocket client for live dashboard updates.
 *
 * Reconnects with exponential backoff — during a cyclone the dashboard is exactly
 * what must not silently stop updating.
 *
 * Event types: cyclone.detected, cyclone.updated, prediction.issued,
 * alert.issued, alert.acknowledged, pipeline.status
 */

const WS_URL = import.meta.env.VITE_WS_URL || `ws://${window.location.host}/ws`;

export class LiveConnection {
  constructor(url = WS_URL) {
    this.url = url;
    this.socket = null;
    this.handlers = new Map();
    this.reconnectDelay = 1000;
    this.maxReconnectDelay = 30000;
    this.shouldReconnect = true;
  }

  connect() {
    // Reset the flag a previous close() set, otherwise reconnection stays
    // disabled for the life of the page once anything closes this singleton.
    this.shouldReconnect = true;
    this.socket = new WebSocket(this.url);

    this.socket.onopen = () => {
      this.reconnectDelay = 1000;
      this.emit('connection.open', {});
    };

    this.socket.onmessage = (event) => {
      try {
        const message = JSON.parse(event.data);
        this.emit(message.type, message.payload);
      } catch {
        // Malformed frame — drop it rather than tearing down a live connection.
      }
    };

    this.socket.onclose = () => {
      this.emit('connection.closed', {});
      if (this.shouldReconnect) {
        setTimeout(() => this.connect(), this.reconnectDelay);
        this.reconnectDelay = Math.min(this.reconnectDelay * 2, this.maxReconnectDelay);
      }
    };
  }

  on(eventType, handler) {
    if (!this.handlers.has(eventType)) this.handlers.set(eventType, new Set());
    this.handlers.get(eventType).add(handler);
    return () => this.handlers.get(eventType)?.delete(handler);
  }

  emit(eventType, payload) {
    this.handlers.get(eventType)?.forEach((handler) => handler(payload));
  }

  close() {
    this.shouldReconnect = false;
    this.socket?.close();
  }
}

export const liveConnection = new LiveConnection();
